core/include/array/array_sorted_read_state.h

```diff
@@ -201,17 +202,14 @@ class ArraySortedReadState {
   /** AIO counter. */
   int aio_cnt_;
 
-  /** The AIO mutex conditions (one for each buffer). */
-  pthread_cond_t aio_cond_[2];
+  /** Condition variables used in AIO. */
+  std::condition_variable aio_cv_[2];
```

Comment:
The algorithm proceeds as follows:

- Compute a tile slab
- Send an AIO for the first tile slab
- Compute a second tile slab
- Send an AIO for the second tile slab
- Wait for first AIO (here is where `aio_cv_` is used)
- Compute a CPU-intensive copy for the first tile slab while the second AIO is working
- Advance the tile slab id in `aio_cv_` and `copy_id_` etc to indicate you will work on the second tile slab
- Send a new AIO for the place of the *first* tile slab
- Wait for second AIO
- Work on the second tile slab while the AIO for the *first* tile slab is working
- etc.

Essentially, we always have two slots/buffers for two tile slabs. While an AIO is fetching one, the CPU works on the received data for the other. And this works interchangeably between the first and second slots (hence the round-robin increase in the id).